src/ServerHandler.cpp

```diff
@@ -1,8 +1,153 @@
+    size_t start = 0U;
+    size_t end = str.find(delimiter);
+    while (end != std::string::npos) {
+        subStrs.push_back(str.substr(start, end - start));
+        start = end + delimiter.length();
+        end = str.find(delimiter, start);
+    }
+    subStrs.push_back(str.substr(start, str.size() - start + 1));
+}
+
+Frame ServerHandler::receiveFrameFromClient() {
+    boost::unique_lock <boost::mutex> lock(_queues.mutexToServer);
+    Frame frame;
+    if (!_queues.framesToServer.empty()) {
+        frame = _queues.framesToServer.front();
+        _queues.framesToServer.pop();
+    }
+    return frame;
+}
+
+void ServerHandler::sendFrameToClient(Frame frame) {
+    boost::unique_lock <boost::mutex> lock(_queues.mutexFromServer);
+    _queues.framesFromServer.push(frame);
+    _queues.condFromServer.notify_all();
+}
+
+void ServerHandler::parseUserFrame(Frame frameFromClient) {
+    FrameCommand cmd = frameFromClient.getCommand();
+    switch (cmd) {
+        case CONNECT: {
+            std::string host = frameFromClient.getHeaderVal("host");
+            short port = boost::lexical_cast<short>(frameFromClient.getHeaderVal("port"));
+            _connectionHandler = new ConnectionHandler(host, port);
+            if (!_connectionHandler->connect()) {
+                Frame frame = Frame(ERROR, "Could not connect to server");
+            }
+            else {
+                _connectionHandler->sendFrameAscii(frameFromClient.toString(), STOMP_DELIMITER.c_str()[0]);
+            }
+            break;
+        }
+        case DISCONNECT: {
+            _connectionHandler->sendFrameAscii(frameFromClient.toString(), STOMP_DELIMITER.c_str()[0]);
+            delete _connectionHandler;
+        }
+        default:
+            _connectionHandler->sendFrameAscii(frameFromClient.toString(), STOMP_DELIMITER.c_str()[0]);
+    }
+}
+
+std::string ServerHandler::extractBookName(std::vector <std::string> pieces, size_t startPos, size_t endPos) {
+    std::string book;
+    // the || is unnecessary with right usage, added here only because the sake of style.
+    for (size_t i = startPos; i < endPos || i < pieces.size(); i++) {
+        book += pieces[i];
+    }
+    return book;
+}
+
+std::string ServerHandler::booksString(std::vector <std::string> books) {
+    std::string body = _inventory.getUsername() + ":";
+    for (std::string book : books) {
+        body += book;
+        body += ",";
+    }
+    body.substr(0, body.size() - 1); // instead of adding a check for last book, just remove last ','
+    return body;
+}
+
+void ServerHandler::parseMessageFrame(Frame messageFrame) {
+    std::string genre = messageFrame.getHeaderVal("destination");
+    std::vector <std::string> subStrs;
+    std::string body = messageFrame.getBody();
+    split(body, subStrs, " ");
+    if (subStrs[1] == "wish") {
+        std::string book = extractBookName(subStrs, 3, subStrs.size()); // starting from the word 'borrow'
```

Comment:
This will include the word borrow in the book name
"Idan wish to borrow Harry Potter"
The word borrow is in index 3 and extractBookName takes the first element.
Also yu can probably get rid of the magical number

Reply:
basicly I would agree about magic number, but:
a. they gave this all magically rubbishly locations.
b. there is so many magic numbers that we won't get out of it...